### ISSUE TITLE
"Risiko-Ermittlung gestoppt" though it had not been turned off

### ISSUE THREAD

**Issue by kereng5** [corona-warn-app/cwa-app-android#3093]
## Describe the bug

![gestoppt_5168](https://user-images.githubusercontent.com/69790073/117271920-f4115b80-ae5a-11eb-9100-23badf84c35d.png)
I did not stop the exposure check, and it was active at 2:46 am. So who stopped it and why? 

One more small issue: When I click "Aktivieren" to reactivate the exposure check, switch the slider to the right and return to the previous screen by the "back triangle", it still shoes "Risiko-Ermittlung gestoppt", and the slider is left again. It works if I switch the slider to the right, close the CWA and start it new.

## Steps to reproduce the issue

I just start the phone in the morning and read "Risiko-Ermittlung gestoppt" in the CWA.
It happened today and on the day before yesterday, but not yesterday.

## Technical details

    Mobile device: Motorola G3
    Android version: 6.0.1
    CWA version 2.1.2
    ENF 18211213000
    no "Priorisierte Hintergrundaktivität"

---
Internal Tracking ID: [EXPOSUREAPP-7095](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7095)

**Comment by dsarkar** [comment]
@kereng5 Thanks for the report. We getting since version 2.1 several of these reports. 
- Can you confirm that you have a stable internet connection?
- Can you confirm that this issue araised only after updating to CWA version 2.1

**Comment by DerVogel2020** [comment]
This is similar to #2880 .

**Comment by dsarkar** [comment]
Internal Tracking ID: [EXPOSUREAPP-7095](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7095)

**Comment by vaubaehn** [comment]
Hi @DerVogel2020 and all, it's not similar to #2880. In my case, a different error occured ("Risiko-Überprüfung fehlgeschlagen" vs. "Risiko-Ermittlung gestoppt"), and contact tracing ("the switch") was still enabled.

I wonder, if this is a problem in ENS which is (correctly) reflected in CWA? (ENS got a hick up, and turned itself off, or something).
@kereng5 how is the memory situation of your internal storage? How much free space is in your internal memory?

**Comment by MikeMcC399** [comment]
@kereng5 
> ENF 18211213000

It may be worth checking if there is an update to the Exposure Notifications System for your device, in case it is an issue with this part of the system.

On an Android 8 device I can do:
Einstellungen > Apps > Google Play-Dienste
tap on App-Details in Store

If there is an update available, it will offer it to you. That may be different on your device / your Android 6.

I'm currently using 21.15.15 for Google Play-Dienste.

In Einstellungen > Google > COVID-19-Benachrichtigungen (scroll down to bottom) this shows as:
Version v1.8 (211515000).

**Comment by dsarkar** [comment]
@kereng5 
Could you provide us with a screenshot please:

Einstellungen -> COVID-19-Benachrichtigungen 

Thanks.

---
Corona-Warn-App Open Source Team

**Comment by kereng5** [comment]
@dsarkar
>     * Can you confirm that you have a stable internet connection?
Yes, the phone is in my WLAN all night,  but sometimes it shows "4G" for the first few seconds.
It is also connected to the charger. Though "Priorisierte Hintergrundaktivität" is not active, there is a check between 2 and 3 almost every day, but not on May 4. The last day before May 4 without a check at night was April 17. This behaviour is not quite deterministic.
>     * Can you confirm that this issue araised only after updating to CWA version 2.1
I have 2.1.2 since May 3, and the issue appeared on May 4 and 6.
The exposure log now shows up to four checks in one day. Before CWA version 2.1.2 there had been only one or two:
[all-exposure-checks06.txt](https://github.com/corona-warn-app/cwa-app-android/files/6435183/all-exposure-checks06.txt)

I don't think a screenshot of COVID-19-Benachrichtigungen will add any useful information. What would you like to see?

**Comment by dsarkar** [comment]
@kereng5 Thanks for the feedback. Developer asked for the screenshot to see if the slider is greyed out.

**Comment by vaubaehn** [comment]
@kereng5 
> The exposure log now shows up to four checks in one day. Before CWA version 2.1.2 there had been only one or two:

This is, because there was/is a bug in OS on how one certain network parameter is determined. Since CWA 2.1.2 it is now worked around that issue, so we Android 6 users get exposure check updates multiple times a day, finally. See: #2881 and #2934.

But looking to your [all-exposure-checks06.txt](https://github.com/corona-warn-app/cwa-app-android/files/6435183/all-exposure-checks06.txt), I see you are/were also affected of #2880, see entries for April 23rd/24th, April 30th/May 1st.

Could you have a look, how much free space there is for your internal storage, please?

**Comment by kereng5** [comment]
@vaubaehn 
There are two entries for April 24 at 02:51 and 12:03 with the same hash and keyCount. Also for May 1, 02:09 and 08:59.
When I noticed cases like that last year, I thought the server data would not change in one day.
Now I am pleased to see that the four entries for May 3 or 5 all have different data.

**Gerätespeicher**
7,37 GB von 8,00 GB insgesamt verwendet
Android Betriebssystem 3,46 GB
Interner Speicher: 3,91 GB von 4,45 GB verwendet

**Arbeitsspeicher**
833 MB von 898 MB, 93%

The screenshot that @dsarkar resp. the devs asked for:
![slider_5169](https://user-images.githubusercontent.com/69790073/117336261-4c1c8200-ae9c-11eb-8fba-6f2519f8850e.png)

**Comment by kereng5** [comment]
Today same as yesterday "Risiko-Ermittlung gestoppt ... Aktualisiert Heute, 02:59".
Tap on "Risiko-Ermittlung aktivieren", move the slider to the right. Tap on "x": still "Risiko-Ermittlung gestoppt".
Move the slider to the right again, tap on  the "back triangle": still "Risiko-Ermittlung gestoppt".
Move the slider to the right again, close the CWA and restart it: exposure check is performed with new server data.

**Comment by dsarkar** [comment]
@kereng5 Thanks for the feedback. Will be forwarded to developers.

**Comment by kereng5** [comment]
@MikeMcC399 I have v1.8 (211515000) now. But I think this update happened after the exposure check.

**Comment by kereng5** [comment]
@MikeMcC399 
> I assume that this was an automatic update if you are not sure when it happened.

When I go to Einstellung > Apps > Google Play-Dienste there is no "Store" and no "App-Details".
And in the Google Play Store I cannot find Google Play-Dienste, so I had to wait for the automatic update.

**Comment by MikeMcC399** [comment]
@kereng5 

> When I go to Einstellung > Apps > Google Play-Dienste there is no "Store" and no "App-Details".

Here is what it looks like on my Android 8 device. It's a pity that function isn't available to you on Android 6.

![image](https://user-images.githubusercontent.com/66998419/117485291-80f60b00-af68-11eb-9b37-da92f7e47e81.png)

> And in the Google Play Store I cannot find Google Play-Dienste, so I had to wait for the automatic update.

The direct link for "Google Play-Dienste" is https://play.google.com/store/apps/details?id=com.google.android.gms and it seems to be hidden from search in Google Play Store. I opened that link in Google Chrome browser on the mobile device and Chrome gave me the option "IN PLAY STORE APP ÖFFNEN". If you try that and it is successful for you, you should just see the option "Deaktivieren" since you already have the latest version of Google Play-Dienste (aka gms). If you didn't have the latest version it would offer to update at that point.

**Comment by kereng5** [comment]
The problem did not occur today. Exposure checks at 02:10 and 08:15.
The reason could be v1.8 (211515000) or switching the phone off for a minute, which I did yesterday. It is much faster now.
**Arbeitsspeicher** 668 MB von 898 MB, 74% (two days ago:  93%)

**Comment by kereng5** [comment]
"Risiko-Ermittlung gestoppt" has not occured again. I had it only three times on May 4, 6, and 7.
Yesterday there was no exposure check at night between 2 and 4, and today it has the timestamp 2:21 in CWA and 2:34 in ENS which means it took 13 minutes (?). Before that 7 minutes was the longest check.

**Comment by dsarkar** [comment]
@kereng5 Thanks for the feedback. 
- Can you confirm that with the ENF update  "Risiko-Ermittlung gestoppt" did not occur again?
- Regarding the duration of the checks, I will consult a developer.

Thanks again. Best, DS

---
Corona-Warn-App Open Source Team

**Comment by thomasaugsten** [comment]
On old phone this duration can happen

**Comment by kereng5** [comment]
@dsarkar 

>     * Can you confirm that with the ENF update  "Risiko-Ermittlung gestoppt" did not occur again?
CWA 2.1.2 came on May 3.
"Risiko-Ermittlung gestoppt" on May 4, 6, and 7.
ENF update v1.8 (211515000) on May 7.
No "Risiko-Ermittlung gestoppt" on May 8, 9, and 10.
I am not quite sure whether "Risiko-Ermittlung gestoppt" on May 7 happened before or after the ENF update, but I guess the update was later.

**Comment by dsarkar** [comment]
@kereng5 ok, thanks for the information. Let us know if the problem persists or disappeared for good in the next few days, please. Thanks.

---
Corona-Warn-App Open Source Team

**Comment by vaubaehn** [comment]
Hi @kereng5 ,
unfortunately I don't have much time at the moment, but I really want to ask you some more diagnostic questions:

When you (try to) re-enable the slider in CWA, do you get any notification/pop-up/dialog, that contact tracing is now turned on (again)?

For more diagnostics for the next time, could you please try following: First, today go to Android > Settings > Google > Covid-19 Benachrichtigungen and stay on this "start page" of Google's Exposure Notification System (ENS). Close the Screen with the round circle at the buttom of the screen. Then, when you open the next time CWA and you see, that the slider is off in CWA, directly jump to the "start page" of the ENS by clicking on the square at the right in the bottom of your screen, chose the ENS page and open it. Is the slider turned off here, too? Or is it still on, contradictory what CWA displays?

Thanks, V.

**Comment by kereng5** [comment]
@vaubaehn
> When you (try to) re-enable the slider in CWA, do you get any notification/pop-up/dialog, that contact tracing is now turned on (again)?

I only got a pop-up telling me that I also should activate "Priorisierte Hintergrundaktivität", which I ignore because I open the CWA every day. The pop-up "Covid-19-Benachrichtigungen aktivieren?" did not occur.

I have seen this pop-up just now fo the first time, when I tested what happens in CWA and in ENS when I turn the Risiko-Ermittlung off and on again. It behaves differently now than it does in the morning when CWA says "Risiko-Ermittlung gestoppt".

**Comment by kereng5** [comment]
Yesterday in the afternoon, when the CWA behaved as it should, the slider in ENS was always following the slider in CWA. Now, in the morning of May 12, the slider in ENS is right in the active position "Covid-19-Benachrichtigungen verwenden", while the slider in CWA is left, inactive, "Risiko-Ermittlung gestoppt". I can move it to the right an click OK on the pop-up "Priorisierte Hintergrundaktivität deaktiviert". The slider is right while the text still says "Risiko-Ermittlung Gestoppt":
![slider_5170](https://user-images.githubusercontent.com/69790073/117949584-8e1d4c00-b312-11eb-80f8-81244ddcdc4c.png)
Sometimes I can see the slider move from right to left, when I return to the CWA, but that is not reliably reproducible. The slider in ENS stays right (on) during all my tests and in Einstellungen-Google it says "Covid-19-Benachrichtigungen an".

I almost never turn the CWA off. It is always in the list of small app screens that I get with the square in the bottom line of my phone. I only turn it off and on again to get out of this "Risiko-Ermittlung Gestoppt" situation. After that CWA says "Risiko-Ermittlung aktiv" and performs an exposure check.

**Comment by vaubaehn** [comment]
@kereng5 Thank you very much for your detailed description! Imho, they are the most valuable and important to get more insight into what caused the problem!

Obviously the link between CWA and ENS is broken in a certain way, otherwise the sliders would be synchronized, but they're not.

I assume, there could be 2 reasons for this issue, theoretically:
1. ENS is crashed, and does not respond to API calls from CWA (i.e., does not provide information, whether tracing is enabled). This might have happened in the context of a currently rolling out ENS update.
2. CWA has a problem. For me this seems to be more likely, as the restarting of CWA sometimes solves the problem.

If CWA has a problem, then there may be either an inconsitency in the persisted data of the tracing status, or for some other reason CWA cannot cleanly determine whether tracing is enabled.

@d4rken Do you think, #3122 may already have an impact on this issue here?
As I could see, tracing status will be logged better in CWA 2.3.x - should @kereng5 enable debug log in the upcoming CWA 2.2 though?

@dsarkar @heinezen 
When you check Google Play Store reviews, you will find a lot of problem reports of this issue for the last days. Could you please verify and consider to forward them to dev department, so that priority of the issue here might be re-evaluated?
Thanks a lot!
See you, V.

**Comment by dsarkar** [comment]
@kereng5  @vaubaehn Thanks for the comments here. This issue has been raised internally and developers are aware of the issue. 

@kereng5 Could you produce a error log? 
https://www.coronawarn.app/de/faq/#error_log
https://www.coronawarn.app/en/faq/#error_log

Thanks.

**Comment by kereng5** [comment]
@dsarkar The FAQ say that I need 2.2 for a error log, and that I should have 2.2.1, but Google Play Store tells me that 2.1.2 is the newest version.

**Comment by DerVogel2020** [comment]
@kereng5 It's the same as for the previous updates. In the blog it's said:
"Version 2.2 - like previous versions - will be delivered in a controlled rollout and is made available for users in waves. While users can manually trigger an update in Apple’s App Store, this option is not available in the Google Play Store. There, the Corona-Warn-App’s new version can be available up to 48 hours later."
https://www.coronawarn.app/en/blog/2021-05-12-corona-warn-app-version-2-2/

**Comment by kereng5** [comment]
I am having the "Risiko-Ermittlung Gestoppt" condition again, for the first time in the afternoon.

Next test: I go to Einstellung/Google/Covid-19-Benachrichtigungen. "Covid-19-Benachrichtigungen verwenden" is on, and I move the slider to the left and click the pop-up away. Now the slider is off in ENS and in CWA. I move it to the right in CWA and get both pop-ups: "Covid-19-Benachrichtigungen aktivieren?" (yes, "aktivieren) and "Priorisierte Hintergrundaktivität deaktiviert" (OK). Now the slider in ENS is right again, but CWA still says "Risiko-Ermittlung Gestoppt" as in my screenshot 6 hours ago, where we can see an inconsistency in CWA.

Next test: I turn the phone off for a minute. After restart of the phone the CWA is still in the list of apps that are shown by clicking the square in the bottom line. The small screen in this list still shows "Risiko-Ermittlung Gestoppt". I tap on it and now it says "Risiko-Ermittlung aktiv" and "Aktualisiert: Heute 16:28". That was just some minutes before I saw "Risiko-Ermittlung Gestoppt".

**Comment by Lu9Bro** [comment]
I have the same problem with my

Device name: Samsung Galaxy S5 mini
Android version: 6.0.1
CWA version 2.1.2 and 2.2.1
ENF 18211515000
"Priorisierte Hintergrundaktivität" ON
16 GB total memory
1.13 GB free memory
WLAN is available continuously.
I switch the device in Offline Mode during the night, and back in Online Mode in the morning.

(I have the 100% same behavior of the CWA app and the slider; I didn’t verify so deeply as @kereng5 did if my ENS behavior is also 100% the same.)
The problem exists with CWA version 2.1.2 and 2.2.1.

While the CWA was trying to download the exposure check data and I was watching the device during this process, I got an „URSACHE 9002 Timed out“ with the following details:
![Screenshot_2021-05-12-07-10-21 crop](https://user-images.githubusercontent.com/84101747/118262195-d4aaac00-b4b4-11eb-9695-f972b1748623.png)

Maybe this helps.
If you need more tests, let me know.

**Comment by Ein-Tim** [comment]
One more Twitter user reporting this Problem: https://twitter.com/pantheraleo356/status/1393239658700681216?s=21

**Comment by kereng5** [comment]
Meanwhile I have Version 2.2.1, but "Risiko-Ermittlung Gestoppt" did not accur in the last three days.

On May 14 "Anzahl der Schlüssel" or keyCount was 376735 at 03:00, but only 20057 at 07:18, 30358 at 17:28, and 35248 at 21:37, which is less than one-tenth of what it should be. The next day it was back in the normal range with 367569 at 02:43. Only on my phone?
Sorry for being off-topic, but this irregularity seems to small to start a new issue.

[all-exposure-checks16.txt](https://github.com/corona-warn-app/cwa-app-android/files/6489516/all-exposure-checks16.txt)

**Comment by MikeMcC399** [comment]
@kereng5 
I don't have any matches for your low numbers on May 14, 2021. They are all 6 figure keyCounts. Your 5 figure ones look suspicious!

timestamp | keyCount
-- | --
16 May 2021, 15:33 | 352445
16 May   2021, 11:24 | 348961
16 May 2021, 07:24 | 347322
16 May   2021, 03:04 | 346709
15 May 2021, 22:42 | 367304
15 May   2021, 18:42 | 364064
15 May 2021, 12:39 | 356761
15 May   2021, 08:39 | 352629
15 May 2021, 04:36 | 351723
15 May   2021, 00:03 | 377459
14 May 2021, 19:32 | 373401
14 May   2021, 15:30 | 367875
14 May 2021, 11:27 | 363025
14 May   2021, 07:23 | 360364
14 May 2021, 03:22 | 359471

**Comment by vaubaehn** [comment]
@kereng5 
I had a similar issue for the days February 4 and 5. @ndegendogo found out, that it was just a single diagnosis key file that was 14 days old at that time (instead of the full package holding the latest 14 days' diagnosis keys). I'm quite sure that this is not related to the issue here.
I think it could either be worth to open a new issue for this, or maybe to extend #2880 - it could be a problem in CWA's caching mechanism of diagnosis key files.
@d4rken FYI: https://covidapps.slack.com/archives/C0194ML0MLN/p1612438180080300?thread_ts=1612438180.080300&cid=C0194ML0MLN

**Comment by ndegendogo** [comment]
@kereng5 I can confirm that your 5-digit numbers are way too low; although I couldn't find fast partial sums of the true set of keys that match your numbers.
Incidentally, it looks as if the server had a hickup / slow response times on that day. At 10:42 I had correctly  8 hourly files (since 00:00 UTC) and 14 daily (the previous 14 days). But at 14:47, it was **only 9 hourly files instead of 12**, plus the expected 14 daily files. At 18:48 it was back to normal (16 hourly and 14 daily files).

**Comment by ndegendogo** [comment]
@kereng5 @vaubaehn wow - meanwhile I succeeded to find the sum! 
@vaubaehn you are on the right track!!!

Your number 20057 at 07:18 is the sum of the 5 hourly files so far (0 + 36 + 169 + 481 + 207 = 893) and the daily file of the previous day (13 May = 19164 keys).

**Comment by ndegendogo** [comment]
And the good news is: although it checked not all files (for whatever reason), at least **the newest files were always included in the check**.

**Comment by dsarkar** [comment]
@kereng5 @Lu9Bro If you could provide us with an error reports, once you could reproduce the issue or the issue occurred again, that might help the developers to investigate the issue further. You need CWA version 2.2 for this. See https://www.coronawarn.app/de/faq/#error_log
https://www.coronawarn.app/en/faq/#error_log

Thanks.

---
Corona-Warn-App Open Source Team

**Comment by vaubaehn** [comment]
@kereng5 and @Lu9Bro : to prevent any misunderstanding regarding @dsarkar 's suggestion - it would mean if you could already turn on the error logging via "CWA > App-Informationen > Fehlerberichte" now, and after the issue occured again, upload the log and provide the ticket numbers here.

**Comment by Lu9Bro** [comment]
Today, I managed to record the issue using the new "error_log" function.
I uploaded the logfile with ID  1C18577E0307D2148D44 .

Unfortunately, I took long time until the error occurred.
It occurred at some point in time between 12:59 and 17:27 (MESZ).
Tomorrow, I'll try to catch it again and within a shorter time range.

**Comment by dsarkar** [comment]
@Lu9Bro Thank you very much. This has been forwarded to the internal ticketing system.
Best wishes, DS

---
Corona-Warn-App Open Source Team

**Comment by dsarkar** [comment]
@kereng5 Thank you very much. Will be forwarded!

---
Corona-Warn-App Open Source Team

**Comment by cydur** [comment]
Similar problem on my Sony Xperia Z3 Compact, Android Version 6.0.1. 
CWA Version 2.2.1
Once a day "Risiko Ermittlung" goes off and cannot be reactivated. It is possible to move the switch to on in the "Risiko Ermittlung aktivieren" page but the main page (screenshot) does not change.
I have to hard kill the app in the Android Settings -> Apps section.
Then after a restart the "Risiko Ermittlung" is back on even without turning the switch.

![Screenshot_20210520-194749](https://user-images.githubusercontent.com/66780478/119026130-1d76cf00-b9a5-11eb-90a4-df07d548bc9f.png)

**Comment by kereng5** [comment]
@dsarkar 
I have the error log running all the time, but the error doesn't happen any more - ten days in a row now (40 successful exposure checks).

**Comment by dsarkar** [comment]
@kereng5 OK, thanks for the feedback. Somehow this is good news. If the log file does not grow to big you might want to continue the logging. Or alternatively, stop and delete the current one and start a new error report.

Thanks.

---
Corona-Warn-App Open Source Team

**Comment by dsarkar** [comment]
@cydur Thanks for your report. Could you please create an error report, see this comment and the following comments, please: 
https://github.com/corona-warn-app/cwa-app-android/issues/3093#issuecomment-843133619

**Comment by Lu9Bro** [comment]
On my device, the problem still occurs 1...3 times per day.
This is a bit less frequent than the weeks before.

**Comment by kereng5** [comment]
@Lu9Bro Yes, I had "Risiko-Ermittlung gestoppt" only when I started the phone in the morning (exept one time in the afternoon). When I start CWA, I often see "Daten werden heruntergeladen", but I never saw "Daten werden heruntergeladen" and then "Risiko-Ermittlung gestoppt".

**Comment by cydur** [comment]
@Lu9Bro Maybe. Since I'm trying to create an error report the problem didn't occur. This might be because I open the app much more frequently to check.

**Comment by kereng5** [comment]
Today at 6 pm, I opened the CWA and it said "Daten werden heruntergeladen". After 15 minutes it showed "Aktualisiert Heute 11:56" again. The exposure log doesn't show the failed attempt, too. Maybe this has nothing to do with this issue.

**Comment by Lu9Bro** [comment]
Since 1 -2 weeks, this error did not occur again.
I don't know if

- it does not occur because the covid-19 incidence value has decreased much,    or
- because the ENF version has automatically updated from 18211515000 to 18212002104 in the meantime.

But now the "Daten werden heruntergeladen" and "Prüfung läuft" work faster then before.
Much faster than expected (proportionally to the decreasing incidence values).
I'm still using CWA version 2.2.1 , just to watch this issue a bit longer.

**Comment by vaubaehn** [comment]
Unfortunately this issue has not be fixed yet. There were some reports since release 2.5.1, here is the latest from Google Play Store:

> Ich war bisher ganz zufrieden. Aber warum ist jetzt auf einmal mein Impf-Zertifikat weg? Auch stellt sich die Risiko Ermittlung immer wieder auf inaktiv und ich kann es nicht wieder aktivieren.

So it's even a combination with #3549 

@d4rken 
Is the switch of the Risiko-Ermittlung also part of a (`HotData`) flow?

**Comment by d4rken** [comment]
> @d4rken
> Is the state of the switch of the Risiko-Ermittlung also part of another (`HotData`) flow?

No: 
https://github.com/corona-warn-app/cwa-app-android/blob/63df24be9d2bc1bfb324325d79b1c8b0f8606f4d/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tracing/DefaultTracingStatus.kt

I don't think there is any direct relation between the two issues, except the device possibly being under a lot of load, causing the ENS communication to timeout/hang, and causing the certificates to hang or more just run into an edgecase more easily.

**Comment by vaubaehn** [comment]
@d4rken 
But it looks like a flow, and the state is emitted?
https://github.com/corona-warn-app/cwa-app-android/blob/63df24be9d2bc1bfb324325d79b1c8b0f8606f4d/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tracing/DefaultTracingStatus.kt#L98-L131

Or is the state requested on other places via `suspend fun` ?

I inspected my logs, and just from the log messages it looks like, the log messages of `isTracingEnabled` from the code above are not really fitting to what happened:
At 06:26 UTC this morning, the watchdog called CWA. DKs were downloaded and checked. The submission to ENF took place in the same minute, but it took 5 minutes, until the task finished:
```2021-07-22T06:41:20.036Z V/TaskControllerExtension: submitBlocking(request=DefaultTaskRequest(type=class de.rki.coronawarnapp.diagnosiskeys.download.DownloadDiagnosisKeysTask (Kotlin reflection is not available), arguments=de.rki.coronawarnapp.diagnosiskeys.download.DownloadDiagnosisKeysTask$Arguments@807aaec, id=66bc7652-ffb9-4731-bbba-0cfae5166cc2, originTag=ExposureWindowRiskWorkScheduler-WatchdogService, errorHandling=null)) continuing with result InternalTaskState(id=ec5d3955-935d-4ecb-97ef-7e3181ad2145, request=DefaultTaskRequest(type=class de.rki.coronawarnapp.diagnosiskeys.download.DownloadDiagnosisKeysTask (Kotlin reflection is not available), arguments=de.rki.coronawarnapp.diagnosiskeys.download.DownloadDiagnosisKeysTask$Arguments@807aaec, id=66bc7652-ffb9-4731-bbba-0cfae5166cc2, originTag=ExposureWindowRiskWorkScheduler-WatchdogService, errorHandling=null), createdAt=2021-07-22T06:36:21.608Z, startedAt=2021-07-22T06:36:21.611Z, finishedAt=2021-07-22T06:41:20.027Z, error=null, result=de.rki.coronawarnapp.diagnosiskeys.download.DownloadDiagnosisKeysTask$Result@c0b2faa, config=Config(executionTimeout=PT480S, collisionBehavior=SKIP_IF_SIBLING_RUNNING), job=LazyDeferredCoroutine{Completed}@bd8b419, task=de.rki.coronawarnapp.diagnosiskeys.download.DownloadDiagnosisKeysTask@67a7cde)```

In the same minute, when the task was started, `isTracingEnabled` was checked, and it failed:
```2021-07-22T06:36:22.219Z I/DefaultDiagnosisKeyProv: Provide diagnosis keys with DiagnosisKeyFileProvider
2021-07-22T06:36:22.221Z D/DefaultDiagnosisKeyProv: Performing key submission.
2021-07-22T06:36:22.219Z I/DefaultENFVersion$isAtL: Comparing current ENF client version 18212418000 with 17000000
2021-07-22T06:36:26.548Z W/DefaultTracingStatus: Quick isEnabled check had timeout, retrying with more patience.
2021-07-22T06:36:26.546Z W/DefaultTracingStatus: Failed to determine tracing status.
kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 1000 ms
	at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:2)
	at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:1)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:21)
	at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:12)
	at java.lang.Thread.run(Thread.java:818)

kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 1000 ms
	at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:2)
	at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:1)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:21)
	at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:12)
	at java.lang.Thread.run(Thread.java:818)
```

I assume, that TimeoutCancellation was not intended in that form? Also there were no more log messages that could have been expected in that context, like
https://github.com/corona-warn-app/cwa-app-android/blob/63df24be9d2bc1bfb324325d79b1c8b0f8606f4d/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tracing/DefaultTracingStatus.kt#L113-L115

Also, what I did not expect, was that new quota was already calculated before `DownloadDiagnosisKeysTask` finished. The quota was already re-calculated at the beginning of the task:
```2021-07-22T06:36:22.176Z D/DefaultENFVersion$requi: Version requirement satisfied: current=18212418000, required=16000000
2021-07-22T06:36:22.180Z D/SubmissionQuota: No new quota available (now=2021-07-22T06:36:22.179Z, availableAt=2021-07-23T00:00:00.000Z)
2021-07-22T06:36:22.180Z D/DefaultDiagnosisKeyProv: Consuming quota: old=5, new=4
```

If the ProvideDiagnosisKeys had been failed, we probably had lost one submission unnecessarily?

Please see also: https://github.com/corona-warn-app/cwa-app-android/pull/3502#issuecomment-865022907

Side note: it does not happen often for me, that the `DownloadDiagnosisKeysTask` takes that long. Not sure, if it was for the first time after many weeks.

**Comment by d4rken** [comment]
> @d4rken
> But it looks like a flow, and the state is emitted?
> https://github.com/corona-warn-app/cwa-app-android/blob/63df24be9d2bc1bfb324325d79b1c8b0f8606f4d/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tracing/DefaultTracingStatus.kt#L98-L131
> 
> Or is the state requested on other places via `suspend fun` ?

I may have misunderstood you then. I meant that it is not a `HotDataFlow` which is a custom `Flow` in the CWA. Yes it's also a `Flow`, just not a `HotDataFlow`.

> I inspected my logs, and just from the log messages it looks like, the log messages of `isTracingEnabled` from the code above are not really fitting to what happened:
> At 06:26 UTC this morning, the watchdog called CWA. DKs were downloaded and checked. 

The watchdog concept is still from before my time at the CWA, and I'm suspicious of it's effectiveness.
Due to background reliability being such a *hottopic*, and there being "theoretically" no complaints (though it may be an unknown unknown), we've not taken taken it to the chopping block...
The idea is that under some circumstances the GPlay services bind to the CWA, and give us guaranteed execution time.
Google didn't explain in detail what is happening there, but it's using something like the service bind mechanism to elevate the CWA background priority to the priority of the play services (and actually start it too).

I think it binds the CWA process for a few seconds, meaning `CoronaWarnApp.onCreate()` executes and after a few seconds unbinds again, and then it's like opening the CWA and pressing the home button again (usual background priority behavior).

These few seconds are enough to schedule our workers again, but the `WatchDogService` actually starts a task right there:
https://github.com/corona-warn-app/cwa-app-android/blob/63df24be9d2bc1bfb324325d79b1c8b0f8606f4d/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/WatchdogService.kt#L50

After play-services unbind us, and if there is nothing else keeping us high-priority (UI or running workers), the system may just kill the CWA process again. The acquired `WakeLock`s, don't do anything to prevent that imho.

This all comes from a time long before we refactored the whole background execution. I don't think it currently helps, but I'm not sure if it **hurts**.

> I assume, that TimeoutCancellation was not intended in that form? Also there were no more log messages that could have been expected in that context, like

The logged exception is a bit misleading, the order of log lines is not in order if you look closely:
* Quick check happens, no answer within `POLLING_DELAY_MS`
* The outer try catch shows the cancelation due to `POLLING_DELAY_MS`
* Then because of that we perform a second check with a longer timeout (`Quick isEnabled check had timeout`)
* If that also times out  you would see `Timeout on ENF side, assuming isEnabled false`

This behavior is not unexpected on slow devices. It's just a cheap way to get a second check if the first one hangs.

> Also, what I did not expect, was that new quota was already calculated before `DownloadDiagnosisKeysTask` finished. The quota was already re-calculated at the beginning of the task:
> 
> ```
> 2021-07-22T06:36:22.180Z D/SubmissionQuota: No new quota available (now=2021-07-22T06:36:22.179Z, availableAt=2021-07-23T00:00:00.000Z)
> 2021-07-22T06:36:22.180Z D/DefaultDiagnosisKeyProv: Consuming quota: old=5, new=4
> ```
> 
> If the ProvideDiagnosisKeys had been failed, we probably had lost one submission unnecessarily?

We don't actually know whether a failed "provideDiagnosisKeys" (if it fails on ENS side) counts against the quota or not, we just assume. In any case the quota calculation was actually never armed, it still just counts.
Exceeding the quota is currently just prevent by scheduling.

> Please see also: [#3502 (comment)](https://github.com/corona-warn-app/cwa-app-android/pull/3502#issuecomment-865022907)
> 
> Side note: it does not happen often for me, that the `DownloadDiagnosisKeysTask` takes that long. Not sure, if it was for the first time after many weeks.

> I think it's a good idea to do this, but I would suggest to do it the other way round: first check 2min timeout, 2nd check 1s. Why?
> It's very plausible that the system as a whole or just the ENF is too busy "with something" that it can't consume the API call or respond within timeout time. For the cases where this is going to happen, it would mean, that the now proposed first check (1s) fails due to system/ENS being busy. The second call will follow immediately after, and there is a high likelihood, that system/ENS is still busy and will also not answer the second call in time.
If you do it the other way round, (2min timeout, then 1s) the likelihood is high, that system/ENS is already recovered when the 2nd call comes in, and can respond.
What do you think?

I don't think the play services throw away the api calls, so if it recovers we should still get an answer within the timeout window of the first call. There may be a chance that the 2nd 1second call succeeds, but if we fail a quick call, and then do a long one (like it is now), the chance to get an answer, if the system is busy, would be equally high. I don't think two requests is enough to DDOS the system, and I don't think play services punish callers for this.
The downside of changing the order, is that if it is any other issue than "system is busy", then the UI has to wait 2 minutes for the second try and would just be blank until then?

> I was also wondering, what can make ENS that busy not to respond in time. For me the most plausible explanation would be heavy work load due to verifiying the signature of diagnosis keys, shortly after provision from CWA to ENS.

Not sure how threading/processes works in the play services, could also be totally unrelated, e.g. an app updating, other apps communicating with play services etc...

> Last question: I guess the 2min timeout is set by CWA?

No by the ENS (or a general gplay services timeout). The CWA would timeout at 7-10min if the call itself was part of background execution (workers capped at 10min, `TaskController` tasks at 7min).

**Comment by vaubaehn** [comment]
Hi @mtwalli , @thomasaugsten and all, the ""Risiko-Ermittlung gestoppt" though it had not been turned off" is back.

In the past weeks I found a low number of reports on Google Play Store, and last night I was experiencing it myself:
At 23:59 CET (22:59 UTC) I checked the ENF logs, found last risk check at 19:56 CET. I jumped to CWA using the app selector ("square"), and found CWA presenting risk calculation switch/card in "stopped" state:
<img src="https://user-images.githubusercontent.com/68376758/144072907-9072373f-f0e3-43ac-acc3-2d8a8895c0ec.png" width="200">
Just took screenshot and jumped back to ENF log:
<img src="https://user-images.githubusercontent.com/68376758/144072902-e1c3b814-1134-4ce7-94f1-3068cf1eb38a.png" width="200">
I checked the number of diagnosis key that have been matched successfully before:
<img src="https://user-images.githubusercontent.com/68376758/144072898-0d76ba40-18d8-4b43-a5a3-f090fdef969d.png" width="200">
I jumped back to CWA using app selector, and found CWA waiting for a response from ENF now:
<img src="https://user-images.githubusercontent.com/68376758/144072893-0e109bbd-12bf-4638-ab03-1776108fcf78.png" width="200">
A short time later, CWA got a response from ENF:
<img src="https://user-images.githubusercontent.com/68376758/144072886-0d2649e0-7ee4-48f8-af10-37b31a211eff.png" width="200">
Matching completed successfully:
<img src="https://user-images.githubusercontent.com/68376758/144072883-0f652893-6a93-4daf-aab2-8b04b6b88dd8.png" width="200"><img src="https://user-images.githubusercontent.com/68376758/144072873-8994bb6b-b433-405d-8700-ff01fe01987b.png" width="200">

Error log was (like always) running in background, and I had a look:
It was clear, that the both timeouts (1-2 check: #3502) timed out, so `tracingEnabled` was emitted as `false`. For some reason, another timeout check was started soon after. This time the quick timeout check failed, but ENF did not time out, so the matching and retrieving results succeeded.
If you want to have a look yourself, you will find the error log with ID `823002ED93931D75AFC8` - interesting part starts at 2021-11-29T22:56:49.977Z in line 86716.

While I'm glad that Exposure Checks recovered without interventions, two questions came up:
- Do we need to be afraid, that ENF will timeout more often with increasing numbers of diagnosis keys, and CWA can't recover so that no risk calculation succeeds anymore, especially on slow devices?
- Is is it possible to fine tune the whole process that users won't see the the "Risiko-Ermittlung gestoppt" anymore? Inceasing quick check timeout won't help I guess, as ENF is the bottle neck here. Is it actually necessary to wait for the ENF to finish, or could it be enough to start risk calculation/displaying results, when the intent from ENF is received?

By the way, I found also a (temporary) problem with statistics: security exception because the download file could not be validated with its public key at 2021-11-29T22:59:26.954Z. How can this happen?

**Comment by thomasaugsten** [comment]
The ENF is sometime too slow and this can happens especially when the app is in the background and device is under load. The ENF is not timing out properly. Often it starts without any result.

In theory the security exception cannot happen. Maybe if the downloaded file is corrupted and he can still unzip it.

**Comment by vaubaehn** [comment]
Hi @mtwalli ,
There is an increasing amount of reports in Google Play Store currently that complain about the same issue here.
Is there anything you can do, e.g. circumvent the GMS timeout?

**Comment by Ein-Tim** [comment]
I'm in contact with a Twitter user who encounters this bug, anything she could do to solve it?

**Comment by dsarkar** [comment]
@Ein-Tim  The user can't activate the exposure logging? He flipped the toggle several times (in OS settings) and it always sets back to "deactivated"?

**Comment by Ein-Tim** [comment]
@dsarkar 

After sole time with active exposure logging, the app shows "Risiko-Ermittlung gestoppt" Toggling exposure logging on/off then, doesn't do anything. Clicking on "Risiko-Ermittlung aktivieren", on the card sometimes works and it works again, but sometimes the app still shows "Risiko-Ermittlung gestoppt".

**Comment by Ein-Tim** [comment]
And another Twitter user reporting this: https://twitter.com/helgol/status/1483534553235050515?s=20

--- 

Device name: Galaxy XCover 3 (SM-G389F)
Android version: Android 6.0.1
CWA version: 2.16.2

---

The main screen shows this: 
![FJjns9FVQAUjS2o](https://user-images.githubusercontent.com/67682506/150373307-b3b28676-ee9f-4db8-b9f1-679a964d661c.png)

And after trying to switch exposure logging on again, the user sees this pop up: 

![FJjntJzVgAIFV72](https://user-images.githubusercontent.com/67682506/150373315-efe1d570-6698-440f-9931-9da8349953ea.png)

**Comment by Ein-Tim** [comment]
The user on Twitter captured the error in an error log with the ID A7011DA8AD880A4BEBDF

**Comment by vaubaehn** [comment]
@thomasaugsten @mtwalli 
I could reproduce stopped Risiko-Ermittlung and recovery: obviously CWA did not receive any feedback from ENF API during some action in background and thus considered the ENF to be turned off. I was not able to turn on Risiko-Ermittlung via the button on top of the home nor via the CWA's settings menu. The best evidence for the missing link to the API was, that there was no ENF version number displayed on the bottom of the CWA's App-Information page (as the ENF version is only displayed when CWA thinks API is connected).
This already happened in the past, and I was able to recover from this state by force closing the app, and restarting, then seeing a white blank page for a minute or so, and then getting back to work.
This time I could recover by turning off bluetooth (there was no visible feedback on the home screen about turned off bluetooth!) and turning it on again. This obviously caused an intent received by CWA that re-created the (work)flow, and the switch of Risiko-Ermittlung automatically switched on again.
I recorded an error log of all these steps, from switching CWA to foreground, finding Risiko-Ermittlung stopped, trying to turn it on again without success, then turning off bluetooth and back on again. The white risk card that informed that Risiko-Ermittlung is stopped turned to the state that download from server failed (which was not the case). After the Risk-Ermittlung turned back to on, the risk card switched to the green state.
Fehlerbericht:
2022.02.04 - 17:37:13
ID 3E445FFE70BAA4906366 

Possible fix: enhance/fix UI/Flow that trying to switch on button for Risiko-Ermittlung actually causes CWA to re-connect to ENF API, or implement a regular task via task scheduler.

**Comment by thomasaugsten** [comment]
We saw this already in error reports that the app receive a timeout from the ENF when it try to figure if the ENF is active or not. 
We are not sure what a good solution is to recover the ENF from this state. Your solution was to manually turn off and on bluetooth.

Normally the app should reconnect to the ENF on every risk calculation.
